riscv_analysis/src/gen/function_annotations.rs

```diff
@@ -5,13 +5,100 @@ use std::{
 };
 
 use crate::{
-    cfg::{Cfg, Function},
-    parser::Register,
-    parser::{Info, JumpLinkType, LabelString, ParserNode, With},
+    cfg::{CfgNode, Cfg, Function},
+    parser::{Info, JumpLinkType, LabelString, ParserNode, Register, With},
     passes::{CfgError, DiagnosticLocation, GenerationPass},
 };
 
+struct MarkData {
+    pub found: HashSet<Register>,
+    pub instructions: Vec<Rc<CfgNode>>,
+    pub returns: Rc<CfgNode>,
+}
+
 pub struct FunctionMarkupPass;
+
+impl FunctionMarkupPass {
+    fn mark_reachable(entry: Rc<CfgNode>, func: Rc<Function>)
+                      -> Result<MarkData, Box<CfgError>> {
+        // Initialize data for graph search
+        let mut queue = vec![entry.clone()];
+        let mut seen = HashSet::new();      // CFG nodes seen during the traversal
+        let mut defs = HashSet::new();      // Registers this function writes to
+        let mut returns = None;             // Return instructions in this function
+
+        // Traverse the CFG for all nodes reachable from the entry point
+        while let Some(node) = queue.pop() {
```

Comment:
In the future, I want to separate CFG traversal logic into its own function. Not necessary for now, but just a note so that you can be prepared for it.